### ISSUE TITLE
`org/eolang/lints/Program.defect` Performance: Slow Execution

### ISSUE THREAD

**Issue by volodya-lombrozo** [objectionary/lints#133]
I have added the usage of `Program.defect` to the `jeo-maven-plugin`, and the plugin now runs three times slower than usual. I conducted performance profiling and found that the `org/eolang/lints/Program.defect` method accounts for 87% of the total execution time
.
Please address the performance issues related to `org/eolang/lints/Program.defect`.

[flamegraph-disassemble-27746.html.txt](https://github.com/user-attachments/files/18092356/flamegraph-disassemble-27746.html.txt)

**Comment by volodya-lombrozo** [comment]
@yegor256 Take a look, please. It's a blocker as well.

**Comment by yegor256** [comment]
@volodya-lombrozo which constructor of `Programs` do you use?

**Comment by yegor256** [comment]
@h1alexbel maybe you can help us investigate this?

**Comment by volodya-lombrozo** [comment]
@yegor256 Please note, this is an urgent issue. It actually blocks any new features in `jeo-maven-plugin`. We need to resolve it as soon as possible.

**Comment by yegor256** [comment]
@volodya-lombrozo I'm afraid, we can't solve it entirely. Some extra performance burden is expected and inevitable. Earlier, we had all linters inside eo-maven-plugin. Now they are here. Obviously, they take time. In your particular case (you may have very large XML files), the performance cost may be very high. The best you can do immediately is simply stop checking all XMIR files via linters, but only check some small files. You need linting simply to validate the correctness of your output -- no need to check all big files, just check the smallest one.

**Comment by volodya-lombrozo** [comment]
@yegor256 It's just faster to fix the performance issue here. Otherwise, I will need to 'invent' a lot of workarounds to hide the original problem, which I will need to remove as soon as a proper solution is implemented.

**Comment by yegor256** [comment]
@volodya-lombrozo I'm not sure there is any "issue" here: there are just 30+ lints which take time when XML files are huge. It's not an issue, it's cost of doing linting :)

**Comment by h1alexbel** [comment]
@yegor256 @volodya-lombrozo how about we add new benchmark test to check `Program.defect` performance?

**Comment by yegor256** [comment]
@h1alexbel not a bad idea to create a benchmark: it should calculate how much it takes to lint a small XMIR, a large, an X-large, an XXL, etc. We can publish this information in README, similar to how we publish it here: https://github.com/objectionary/hone-maven-plugin/blob/master/.github/workflows/benchmark.yml

**Comment by h1alexbel** [comment]
@yegor256 assign me, I will try to

**Comment by yegor256** [comment]
@h1alexbel take it, in #134

**Comment by volodya-lombrozo** [comment]
I've provided one more profiling round for the version `0.0.27`. Here are results:
[flamegraph-disassemble-56062.html.txt](https://github.com/user-attachments/files/18265576/flamegraph-disassemble-56062.html.txt)

As we can see, the `objectionary/lints` still takes unreasonable amount of time

**Comment by VasilevNStas** [comment]
@yegor256 @volodya-lombrozo @maxonfjvipon, this performance report targets `Program.defect()` (87% of execution time). That method and the entire whole-program-analysis layer (`Program`, `Programs`, `WpaLints`, `PkWpa`, etc.) were **removed** in commit `235cda2d` (PR #845, Apr 24, 2026). With the WPA layer gone, the reported hotspot can no longer exist on `master`.

If the concern was about the per-file linting path, the current performance focus shifted to the XSL-based lints — see the XSL speed cluster (#979, #980, #981, #982, #983, #984, #986-#993), where each transformation exceeds the 100ms threshold on large XMIR. I'd suggest closing this one as resolved-by-refactoring and tracking the remaining performance work there.

**Comment by volodya-lombrozo** [comment]
@VasilevNStas Yes, indeed. This issue is outdated and can be closed. Thank you. Btw, we fixed most of the performance issue in other PRs
